### PR TITLE
HyPINO Hypernetwork: per-sample SRF weight generation from flow params

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Generate per-sample SRF MLP weights from global flow parameters (Re, AoA, gap, stagger) using a low-rank hypernetwork. The current SRF uses the same weights for ALL samples, relying only on affine AdaLN conditioning to adapt to different flow conditions. A hypernetwork can produce *completely different pressure field computations* for different tandem configurations — exactly what p_tan needs, since gap/stagger are the primary axes of tandem variance.

**Why this targets p_tan:** The tandem transfer metric (p_tan=26.319) is 2x worse than p_in (11.872). The main source of p_tan error is that aft-foil pressure distributions vary dramatically with gap and stagger ratio, but the current SRF has no mechanism to adapt its function to these parameters beyond scale/shift conditioning. A hypernetwork that sees [Re, AoA, gap, stagger, is_tandem] can generate sample-specific weight corrections, enabling the SRF to compute genuinely different transformations per configuration.

**Literature:** HyPINO (Zhao et al., NeurIPS 2025 Spotlight, ETH Zurich) — "Hypernetwork for Physics-Informed Neural Operators." Also: Cho et al. 2023, "Hypernetwork-based Meta-Learning for Low-Rank Physics-Informed Adaptation."

### Architecture

Add `--hyper_srf` flag. When enabled:

1. **Hyper-encoder:** Small MLP that takes global flow condition vector and outputs low-rank weight corrections for the SRF.
   ```python
   class HyperEncoder(nn.Module):
       def __init__(self, cond_dim=5, hidden=64, rank=8, n_srf_layers=2, srf_hidden=128):
           super().__init__()
           # Total output: for each SRF layer, produce U (srf_hidden x rank) and V (rank x srf_hidden)
           out_dim = n_srf_layers * 2 * srf_hidden * rank  # U + V for each layer
           self.net = nn.Sequential(
               nn.Linear(cond_dim, hidden),
               nn.GELU(),
               nn.Linear(hidden, out_dim),
           )
           # Initialize output near zero so base weights dominate at start
           nn.init.zeros_(self.net[-1].weight)
           nn.init.zeros_(self.net[-1].bias)
           self.rank = rank
           self.n_layers = n_srf_layers
           self.srf_hidden = srf_hidden
       
       def forward(self, cond):
           """cond: [B, 5] = [Re_norm, AoA_norm, gap_norm, stagger_norm, is_tandem]"""
           raw = self.net(cond)  # [B, n_layers * 2 * srf_hidden * rank]
           # Reshape into U, V for each layer
           uv = raw.view(-1, self.n_layers, 2, self.srf_hidden, self.rank)
           return uv  # [B, n_layers, 2, srf_hidden, rank]
   ```

2. **Dynamic SRF:** Each SRF linear layer's weight becomes `W_eff = W_base + U @ V.T` where U, V come from the hyper-encoder:
   ```python
   def dynamic_linear(x, W_base, b_base, U, V):
       """x: [N, in_dim], W_base: [out_dim, in_dim], U: [out_dim, rank], V: [rank, in_dim]"""
       # Base computation + low-rank correction
       h = F.linear(x, W_base, b_base) + (x @ V.T) @ U.T
       return h
   ```

3. **Conditioning vector:** Normalize the 5 scalars to roughly [-1, 1]:
   - Re: `(log(Re) - mean_log_re) / std_log_re` (compute from training data stats)
   - AoA: `AoA / 15.0` (rough normalization)
   - gap: `gap / max_gap` (from dataset)
   - stagger: `stagger / max_stagger`
   - is_tandem: 0 or 1
   
   These should be available in the batch data. If gap/stagger are not directly available, extract from the mesh geometry (aft foil center - fore foil center).

### Training Setup

Use exact baseline hyperparameters with `--hyper_srf`:

```bash
cd cfd_tandemfoil && python train.py \
  --agent tanjiro --wandb_name "tanjiro/hypino-srf-s42" \
  --wandb_group hypino-srf --seed 42 \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --hyper_srf
```

Run both seeds (42, 73).

### Key Implementation Details

- **Rank 8** is enough — the SRF hidden dim is 128, so rank 8 means the correction has 128/8 = 16x fewer params than full rank.
- **Zero initialization** of the hyper-encoder output layer is CRITICAL — ensures the model starts from the baseline weights and only adds corrections as it learns.
- **Only apply to SRF** layers, NOT the backbone. The backbone is shared and should not be sample-specific.
- **For single-foil samples**, set gap=0, stagger=0, is_tandem=0 in the conditioning vector. The hyper-encoder should learn to produce zero/near-zero corrections for these, recovering the baseline behavior.
- **Parameter count:** ~200K extra params (hyper-encoder) — trivial vs 96GB VRAM.

### What to Watch

- **p_tan** is the primary target — gap/stagger adaptation should improve aft-foil predictions
- **p_in** must NOT regress — the zero-init ensures baseline behavior for single-foil
- Monitor hyper-encoder output norms — if they grow large, the corrections may destabilize training
- If p_re improves, that confirms the hypernetwork helps with Re extrapolation too

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.872** | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| **p_tan** | **26.319** | < 26.319 |
| **p_re** | **6.229** | < 6.229 |

Baseline PR: #2357. W&B: aycq1m8m (s42), 9sk276v6 (s73).

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### Summary

HyperSRF did **not** improve on baseline. The primary target (p_tan) regressed slightly. Root cause: an initialization bug caused the hypernetwork to produce zero corrections throughout all 154 epochs — the model effectively reduced to standard SRF.

---

### Metrics

| Split | Metric | Seed 42 (ndw09tm7) | Seed 73 (b9ez9x2x) | 2-seed avg | Baseline | Delta |
|-------|--------|-------------------|-------------------|-----------|----------|-------|
| val_in_dist | mae_surf_p | 11.913 | 11.918 | **11.916** | 11.872 | +0.4% ↑ |
| val_in_dist | mae_surf_Ux | 2.885 | 2.754 | 2.819 | — | — |
| val_in_dist | mae_surf_Uy | 0.834 | 0.849 | 0.841 | — | — |
| val_ood_cond | mae_surf_p | 7.309 | 7.887 | **7.598** | 7.459 | +1.9% ↑ |
| val_ood_cond | mae_surf_Ux | 2.184 | 2.161 | 2.172 | — | — |
| val_ood_cond | mae_surf_Uy | 0.661 | 0.645 | 0.653 | — | — |
| val_tandem_transfer | mae_surf_p | 26.771 | 27.392 | **27.082** | 26.319 | +2.9% ↑ |
| val_tandem_transfer | mae_surf_Ux | 4.275 | 4.145 | 4.210 | — | — |
| val_tandem_transfer | mae_surf_Uy | 1.596 | 1.553 | 1.575 | — | — |
| val_ood_re | mae_surf_p | 6.283 | 6.100 | **6.192** | 6.229 | -0.6% ↓ |
| val_ood_re | mae_surf_Ux | 1.855 | 1.776 | 1.815 | — | — |
| val_ood_re | mae_surf_Uy | 0.453 | 0.433 | 0.443 | — | — |

**All targets missed.** p_tan (primary target) regressed from 26.319 → 27.082 (+2.9%).

---

### Root Cause: Initialization Bug

The hypernetwork outputs [U; V] from a single zero-initialized final layer. This creates a zero-gradient saddle point:
- At init: U=0, V=0, so correction = x @ V @ U.T = 0  ✓ (intended)
- Gradient to U ∝ V.T @ x — **zero because V=0**
- Gradient to V ∝ x.T @ (∂L/∂corr) @ U — **zero because U=0**

The hyper_net parameters receive zero gradients at init and **never escape** — confirmed by W&B logging  and  throughout all 154 epochs. The model trained entirely through the base SRF path, ignoring the hypernetwork.

**Correct LoRA initialization:** Only zero-init U (like LoRA's B matrix), leave V with small random values (like LoRA's A matrix). This preserves zero corrections at init while allowing gradient flow:


With this fix, V is non-zero → gradient flows to U → corrections grow → V also gets gradients. The hypernetwork can actually learn.

---

### Secondary: Verification NaN Bug

The post-training verification sanity check reported NaN for both seeds:

This is a separate bug in the verification loop (the HyperSRF conditioning vector wasn't being passed correctly in the verify_split context). The training/validation metrics are **not** affected — W&B logged valid, non-NaN values throughout. This verification bug would need to be fixed in the verification code path.

---

### Experiment Details

- **Epochs:** 154 (both seeds)  
- **Runtime:** ~210 minutes per run (exceeded 180-min timeout — models completed the running epoch before stopping)
- **Peak VRAM:** 40.5 GB (s42), 40.9 GB (s73)
- **W&B:** [ndw09tm7](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ndw09tm7) (s42), [b9ez9x2x](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/b9ez9x2x) (s73)

**Train command (s42):**
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=1 SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=500 python train.py \
  --agent tanjiro --wandb_name "tanjiro/hypino-srf-s42" --wandb_group hypino-srf \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 --hyper_srf
```

---

### What Happened

The hypernetwork architecture was implemented correctly structurally, but the **zero-initialization of both U and V** from a single output layer created a gradient dead-zone. This is a well-known pitfall with LoRA-style initialization: you must initialize A (V here) randomly and B (U here) to zero — not both to zero. With V=0 at init, the gradient to U through the correction path is V.T @ x = 0, so neither matrix ever updates.

The model was effectively identical to the standard SRF baseline from epoch 1 onward. The slight regression in p_tan/p_oodc likely comes from the additional LayerNorm layers in HyperSRF's base path compared to the original SRF, changing the effective architecture slightly.

---

### Suggested Follow-ups

1. **Fix the initialization and retry (v2):** Use asymmetric LoRA init (U=0, V~N(0, 0.01)). This is a one-line fix that should make the hypernetwork actually learn. With proper gradients, the gap/stagger conditioning could genuinely help p_tan.

2. **Fix verification NaN:** Pass the HyperSRF conditioning vector in the verification loop — currently the verify_split path doesn't capture it before normalization.

3. **Alternative: Additive cond instead of weight generation:** If LoRA-style corrections prove hard to optimize, a simpler approach is to add the conditioning to the SRF hidden states directly (feature-space conditioning rather than weight-space conditioning). More stable to train, though less expressive.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2385 because ADVISOR: Closing — initialization bug means this experiment was essentially standard SRF. Both U and V matrices zero-initialized → zero gradients → hypernetwork never updated. Tanjiro's root cause analysis is spot-on: asymmetric LoRA init (U=0, V~N(0,0.01)) would fix this. However, we're pivoting to bolder experiments per research directive. Tanjiro is already working on #2395 (ANP Re-Conditional AdaLN). The hypernetwork idea could be revisited in a future round with proper init if the concept still seems promising..
